pipeline/data_store.py

```diff
@@ -0,0 +1,773 @@
+import sqlalchemy as sa
+
+from pipeline.utils import get_git_hash, get_latest_provenance
+
+from models.base import SmartSession
+from models.provenance import CodeHash, CodeVersion, Provenance
+from models.exposure import Exposure
+from models.image import Image
+from models.source_list import SourceList
+from models.world_coordinates import WorldCoordinates
+from models.zero_point import ZeroPoint
+from models.cutouts import Cutouts
+from models.measurements import Measurements
+
+
+UPSTREAM_NAMES = {
+    'preprocessing': [],
+    'extraction': ['preprocessing'],
+    'astrometry': ['extraction'],
+    'calibration': ['extraction', 'astrometry'],
+    'subtraction': ['preprocessing', 'extraction', 'astrometry', 'calibration'],
+    'detection': ['subtraction'],
+    'cutting': ['detection'],
+    'measurement': ['detection', 'calibration'],
+}
+
+UPSTREAM_OBJECTS = {
+    'preprocessing': 'image',
+    'coaddition': 'image',
+    'extraction': 'sources',
+    'astrometry': 'wcs',
+    'calibration': 'zp',
+    'subtraction': 'sub_image',
+    'detection': 'detections',
+    'cutting': 'cutouts',
+    'measurement': 'measurements',
+}
+
+
+class DataStore:
+    """
+    Create this object to parse user inputs and identify which data products need
+    to be fetched from the database, and keep a cached version of the products for
+    use downstream in the pipeline.
+    """
+    @staticmethod
+    def from_args(*args, **kwargs):
+        """
+        Create a DataStore object from the given arguments.
+        """
+        if len(args) == 0:
+            raise ValueError('No arguments given to DataStore constructor!')
+        if len(args) == 1 and isinstance(args[0], DataStore):
+            return args[0]
+        else:
+            return DataStore(*args, **kwargs)
+
+    def __init__(self, *args, **kwargs):
+        """
+        See the parse_args method for details on how to initialize this object.
+        """
+        # these are data products that can be cached in the store
+        self.exposure = None  # single image, entire focal plane
```

Comment:
It makes sense to cache the db object, but it does not make sense to cache the pixel values for the exposure.  Usually, the only thing you do with an exposure is strip an image out of it, and each run of the pipeline is going to be working on just one image out of the exposure.  No need to carry around all that extra data.

Reply:
I suppose when we add the code that cuts out the specific CCD image, it will also issue a command to the Exposure object to clear it's internal memory.

Reply:
We shouldn't read the entire exposure into memory in the first place.  For 60 2048x4096 chips, that's 2GB of memory, or 4GB if read as doubles.  If then 60 processes are all reading that, we're going to get into trouble on lots of machines.  But, also, even compressed, the exposure file is half a gig; that's a lot of gratuitous I/O.

We should only read the data for the chip that we're actually processing off of the disk, not the whole thing.

Reply:
OK I can figure that out when we actually build the `Exposure` class.